util/webhook/certs/certmanager.go

```diff
@@ -0,0 +1,275 @@
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"knative.dev/pkg/webhook/certificates/resources"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// New returns a new instance of a CertificateManager.
+func New(options Options) (*CertificateManager, error) {
+	if err := options.defaultOpts(); err != nil {
+		return nil, err
+	}
+	return &CertificateManager{opts: options}, nil
+}
+
+// CertificateManager creates and rotates certificates required by a controller manager's webhook server.
+type CertificateManager struct {
+	opts Options
+}
+
+// Start starts certificate management for a controller manager's webhooks.
+// This method calls os.Exit when an error is encountered.
+func (cm *CertificateManager) Start(ctx context.Context) error {
+	go func() {
+		if err := cm.start(ctx); err != nil {
+			cm.opts.Logger.Error(err, "unexpected failure in starting certificate manager")
+			os.Exit(1)
+		}
+	}()
+	return nil
+}
+
+// start does the actual work of cert rotation.
+func (cm *CertificateManager) start(ctx context.Context) error {
+	var rotationTime time.Time
+	for {
+		nextRotationTime, err := cm.rotateCerts(ctx, rotationTime)
+		if err != nil {
+			return err
+		}
+		rotationTime = nextRotationTime
+		select {
+		case <-time.After(time.Until(nextRotationTime)):
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// rotateCerts rotates certificates at the scheduled rotation time, writes them to the secret and returns the next
+// scheduled rotation time.
+func (cm *CertificateManager) rotateCerts(ctx context.Context, scheduledNextRotationTime time.Time) (time.Time, error) {
+	now := time.Now()
+	cm.opts.Logger.Info("Rotating certificates", "now", now.String(), "scheduledNextRotationTime", scheduledNextRotationTime.String())
+
+	// Get the webhook server's secret. Certificate manager expects an empty secret to have been created with the
+	// controller deployment that intends to use cert management.
+	// Any error in getting that secret is cause for exiting this program.
+	secret := &corev1.Secret{}
+	secretKey := client.ObjectKey{Name: cm.opts.SecretName, Namespace: cm.opts.SecretNamespace}
+	cm.opts.Logger.Info("Getting secret", "namespacedName", secretKey.String())
+	if err := cm.opts.Client.Get(ctx, secretKey, secret); err != nil {
+		return time.Time{}, err
+	}
+
+	if len(secret.Annotations) == 0 {
+		secret.Annotations = map[string]string{}
+	}
+
+	// If this is the first time rotate is called since starting certificate manager, then initialize the webhooks with
+	// any CA data (if any) in the secret. This is for the case when a package is redeployed (for any reason) and the
+	// webhook config properties are reset.
+	if scheduledNextRotationTime.IsZero() {
+		if caCertData := secret.Data[CACertName]; len(caCertData) > 0 {
+			if err := cm.updateWebhookConfigs(ctx, caCertData); err != nil {
```

Comment:
Is this okay to blindly update the webhook config with the cert-data? What if the contents of the secret has proper fields defined but not a valid certificate? What is the certificate present has expired?

Reply:
The chances of expiry are low between rotation intervals and restarts. But maybe an additional validity check here is useful. Let me see how to do that.

Reply:
Thinking about this more, I don't believe it's necessary to add checks here. Here's why:

1. We don't expect anyone to manually edit the webhook secret. So we can be fairly sure that whatever cert is in there was created by certificate manager.
2. Because of 1), the webhook secret will always contains the next rotation time annotation. Even if the certificate has expired, the subsequent code will check the next rotation and rotate the certs.